flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskExecutorManager.java

```diff
@@ -301,6 +307,45 @@ public Map<WorkerResourceSpec, Integer> getRequiredWorkers() {
                 : Collections.emptyMap();
     }
 
+    private Collection<ResourceDeclaration> getResourceDeclaration() {
+        final int pendingWorkerNum =
+                MathUtils.divideRoundUp(getNumberPendingTaskManagerSlots(), numSlotsPerWorker);
+        Set<InstanceID> neededRegisteredWorkers = new HashSet<>(taskManagerRegistrations.keySet());
+        neededRegisteredWorkers.removeAll(unWantedWorkers);
+        final int totalWorkerNum = pendingWorkerNum + neededRegisteredWorkers.size();
+
+        return Collections.singleton(
+                new ResourceDeclaration(
+                        defaultWorkerResourceSpec, totalWorkerNum, new HashSet<>(unWantedWorkers)));
+    }
+
+    private void declareNeededResourcesWithDelay() {
```

Comment:
The `xxxWithDelay` pattern has been repeated 4 times (declarative/fine-grained slot manager, check resource requirement / declare needed resources), and there could be more in future. I wonder if it can be deduplicated as a util. Not necessary for this PR though. This PR is already quite complicated.